mlir/test/Dialect/Mesh/canonicalization.mlir

```diff
@@ -0,0 +1,72 @@
+// RUN: mlir-opt --canonicalize %s | FileCheck %s
+
+mesh.cluster @mesh0(rank = 2, dim_sizes = [2, 4])
+
+// CHECK-LABEL: func @all_reduce_mesh_axes
+func.func @all_reduce_mesh_axes(
+    %arg0 : tensor<4xf32>) -> tensor<4xf64> {
+// CHECK: mesh_axes = array<i16: 0, 1>
+  %0 = mesh.all_reduce %arg0 {
+    mesh = @mesh0, mesh_axes = array<i16: 1, 0, 0>, reduction = #mesh.partial<sum>
+    } : tensor<4xf32> -> tensor<4xf64>
+  return %0 : tensor<4xf64>
+}
+
+// CHECK-LABEL: func @all_reduce_empty_mesh_axes_and_default_reduction
```

Comment:
If `mesh_axes` is empty, do we need to remove the all_reduce op?

Reply:
An empty `mesh_axes` should mean each device in its own group. I think it should be valid. It would have to be optimized out. If we have the rewrite for the empty case then all other passes can benefit and not that it should be a no op.

Reply:
Should this be in the canonicalizer?

Reply:
I think so. We can opt out all the ccl ops when it's empty except all_reduce/reduce_scatter's input and output dtypes are different.

Reply:
Done. Added removal of ops with empty `mesh_axes`.
There could be more optimizations where the device group size ends up being 1, but I have not added that check.

Reply:
It's fine, we can add it later.